src/ledctl/ledctl.c

```diff
@@ -541,87 +589,279 @@ static led_status_t _cmdline_ibpi_parse(int argc, char *argv[])
+ * @param[in]      req            structure with request.
+ *
+ * @return bool if successful, otherwise false.
+ */
+bool _cmdline_parse_params(int opt, int opt_index, struct option *longopts, struct request *req)
+{
+	switch (opt) {
+	int log_level;
+
+	/* Parse major params */
+	case 0:
+		switch (get_option_id(longopts[opt_index].name)) {
+		case OPT_LOG_LEVEL:
+			log_level = get_option_id(optarg);
+			if (log_level != -1)
+				set_verbose_level(&conf, log_level);
+			else
+				return false;
+			break;
+		default:
+			set_verbose_level(&conf, possible_params[opt_index]);
 		}
-	} while (opt >= 0);
+		break;
+	case 'l':
+		set_log_path(&conf, optarg);
+		break;
+	case 'x':
+		listed_only = 1;
+		break;
+#ifdef ENABLE_TEST
+	case 'T':
+		test_params = true;
+		break;
+#endif
+	/* Parse slots params */
+	case 'c':
+		req->cntrl = led_string_to_cntrl_type(optarg);
+		break;
+	case 's':
+	{
+		struct ibpi_state *state = _ibpi_state_get(optarg);
 
-	return status;
+		if (state)
+			req->state = state->ibpi;
+		free(state);
+		break;
+	}
+	case 'd':
+		strncpy(req->device, optarg, PATH_MAX - 1);
+		break;
+	case 'p':
+		strncpy(req->slot, optarg, PATH_MAX - 1);
+		break;
+	case ':':
+	case '?':
+	default:
+		return false;
+	}
+
+	return true;
 }
 
 /**
- * @brief Inits slot request structure with initial values.
+ * @brief Command line parser - options.
  *
- * @param[in]       slot_req       structure with slot request
+ * This is internal function of ledctl utility. The function parses options of
+ * ledctl application. Refer to ledctl help in order to get more information
+ * about ledctl command line options.
  *
- * @return This function does not return a value.
+ * @param[in]      argc           number of elements in argv array.
+ * @param[in]      argv           command line arguments.
+ * @param[in]      req            structure with request.
+ *
+ * @return LED_STATUS_SUCCESS if successful, otherwise led_status_t.
  */
-static void slot_request_init(struct slot_request *slot_req)
+led_status_t _cmdline_parse(int argc, char *argv[], struct request *req)
 {
-	memset(slot_req, 0, sizeof(struct slot_request));
+	int _last_optind, opt, opt_index = -1;
+	char *shortopts;
+	struct option *longopts;
 
-	slot_req->chosen_opt = OPT_NULL_ELEMENT;
-	slot_req->state = LED_IBPI_PATTERN_UNKNOWN;
+	_setup_mode_options(req, &shortopts, &longopts);
+
+	do {
+		_last_optind = optind;
+		opt = getopt_long(argc, argv, shortopts, longopts, &opt_index);
+
+		if (opt == -1)
+			break;
+
+		if (opt == '?' && req->chosen_opt != OPT_IBPI_MODE) {
+			log_error("Parameter %c not recognized, it may be invalid or not correct for this mode.\n",
+				   optopt);
+			return LED_STATUS_CMDLINE_ERROR;
+		}
+
+		if (!_cmdline_parse_params(opt, opt_index, longopts, req)) {
+			log_error("Cannot parse parameter '%s'.", argv[_last_optind]);
+			free(longopts);
+			free(shortopts);
+			return LED_STATUS_CMDLINE_ERROR;
+		}
+		opt_index = -1;
```

Comment:
do we need to invalidate it in every loop?

Reply:
I think no, it should be deleted now.